prov/shm/src/smr_progress.c

```diff
@@ -120,7 +120,7 @@ static int smr_progress_resp_entry(struct smr_ep *ep, struct smr_resp *resp,
 				if (pending->bytes_done != pending->cmd.msg.hdr.size) {
 					FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
 						"Incomplete copy from mmapped file\n");
-					*err = (pending->bytes_done < 0)
+					*err = ((ssize_t) pending->bytes_done < 0)
```

Comment:
I am not familiar with the full context. It seems `pending->bytes_done` is return of `ofi_copy_to_hmem_iov`?

So `ofi_copy_to_hmem_iov`  can return  a value "< 0" under certain  situation?

Reply:
Yeah, `ofi_copy_to_hmem_iov` returns a negative value in error cases. While `ofi_copy_to_iov` returned a size_t, the hmem variant returns a ssize_t. Changing this to a draft because I think there may be other type conversion issues lurking here.

Reply:
In that case, the return value of `ofi_hmem_copy_iov` should be stored to an variable `ret`, and only if `ret` is < 0, we set error. if `ret >=0`, we update bytes_done and proceed.

Reply:
I should have been more specific, the other type conversion issue here is because of the conversion to an int if the conditional is true:
```
((ssize_t) *total_len < 0) ? ((int) *total_len) : -FI_ETRUNC;
```

Type conversion from an unsigned 64 bit to a signed 32 bit integer is undefined behavior in C.

Reply:
Let me know how it looks now. I didn't conditionally set `*total_len` where that is used because it is unused when an error occurs, but I did conditionally set `pending->bytes_done` if no error occurred.